android/java/org/chromium/chrome/browser/search_engines/settings/BraveSearchEngineAdapter.java

```diff
@@ -26,46 +27,66 @@ public class BraveSearchEngineAdapter extends SearchEngineAdapter {
     public static final String STANDARD_DSE_SHORTNAME = "standard_dse_shortname";
 
     private boolean mIsPrivate;
+    private Profile mProfile;
+    private boolean needUpdateActiveDSE;
 
-    static public void setDSEPrefs(TemplateUrl templateUrl, boolean isPrivate) {
+    public BraveSearchEngineAdapter(Context context, boolean isPrivate) {
+        super(context);
+        mIsPrivate = isPrivate;
+
+        // Only need last used profile because we are in settings
+        if (mProfile == null) {
+            if (!mIsPrivate) {
+                mProfile = Profile.getLastUsedRegularProfile().getOriginalProfile();
+            } else {
+                mProfile = Profile.getLastUsedRegularProfile().getPrimaryOTRProfile(/* createIfNeeded= */ true);
+            }
+        }
+    }
+
+    static public void setDSEPrefs(TemplateUrl templateUrl, Profile profile) {
         SharedPreferences.Editor sharedPreferencesEditor =
                 ContextUtils.getAppSharedPreferences().edit();
         sharedPreferencesEditor.putString(
-                isPrivate ? PRIVATE_DSE_SHORTNAME : STANDARD_DSE_SHORTNAME,
+                profile.isOffTheRecord() ? PRIVATE_DSE_SHORTNAME : STANDARD_DSE_SHORTNAME,
                 templateUrl.getShortName());
         sharedPreferencesEditor.apply();
     }
 
-    static public void updateActiveDSE(boolean isPrivate) {
-        TemplateUrl templateUrl = getTemplateUrlByShortName(getDSEShortName(isPrivate));
+    static public void updateActiveDSE(Profile profile) {
+        String shortName = getDSEShortName(profile, false);
+        TemplateUrl templateUrl = getTemplateUrlByShortName(profile, shortName);
         if (templateUrl == null) {
             return;
         }
         String keyword = templateUrl.getKeyword();
-        TemplateUrlServiceFactory.get().setSearchEngine(keyword);
+        BraveTemplateUrlServiceFactory.getForProfile(profile).setSearchEngine(keyword);
     }
 
-    static public String getDSEShortName(boolean isPrivate) {
+    static public String getDSEShortName(Profile profile, boolean javaOnly) {
```

Comment:
Could you explain this `javaOnly` parameter? It's not obvious when it should be `false` and when `true`.

Reply:
I renamed it to `readJavaPrefOnly` and added a comment. Basically when it is `true` the native service will not be used and only Java side preference will be used.